shark/src/main/java/shark/internal/PathFinder.kt

```diff
@@ -462,6 +473,51 @@ internal class PathFinder(
     }
   }
 
+  private fun HeapInstance.readAllNonNullFieldsOfReferenceType(): MutableList<HeapField> {
+    val hprofGraph = graph // DO NOT REMOVE! Getter performs typecast which is expensive!
+    var fieldReader: FieldValuesReader? = null
+    val result = mutableListOf<HeapField>()
+    var skipBytesCount = 0
+    for (heapClass in instanceClass.classHierarchy.toList()) {
+      for (fieldRecord in heapClass.readRecord().fields) {
+        if (fieldRecord.type != PrimitiveType.REFERENCE_HPROF_TYPE) {
+          // Skip all fields that are not references. Track how many bytes to skip
+          skipBytesCount += hprofGraph.getRecordSize(fieldRecord)
+        } else {
+          // Initialize field reader if it's not yet initialized. Replaces `lazy` without synchronization
+          if (fieldReader == null) {
+            fieldReader = hprofGraph.createFieldValuesReader(readRecord())
+          }
+
+          // Skip the accumulated bytes offset
+          fieldReader.skipBytes(skipBytesCount)
+          skipBytesCount = 0
+
+          val fieldValue = ReferenceHolder(fieldReader.readId())
```

Comment:
Is `readId()` really needed here? What's the impact of `readValue()` on the overall time vs readId()? Looks like a micro optimization?

Reply:
If we use `readValue()` it will involve additional cast check, I'm not sure how expensive it is:

`val fieldValue = fieldReader.readValue(fieldRecord) as ReferenceHolder`
```
L20
 LINENUMBER 496 L20
 ALOAD 3
 ALOAD 8
 INVOKEVIRTUAL shark/FieldValuesReader.readValue (Lshark/HprofRecord$HeapDumpRecord$ObjectRecord$ClassDumpRecord$FieldRecord;)Lshark/ValueHolder;
 DUP
 IFNONNULL L21
 NEW kotlin/TypeCastException
 DUP
 LDC "null cannot be cast to non-null type shark.ValueHolder.ReferenceHolder"
 INVOKESPECIAL kotlin/TypeCastException.<init> (Ljava/lang/String;)V
 ATHROW
L21
 CHECKCAST shark/ValueHolder$ReferenceHolder
 ASTORE 10
```
Compared to:
`val fieldValue2 = ReferenceHolder(fieldReader.readId())`

```
L22
 LINENUMBER 498 L22
 NEW shark/ValueHolder$ReferenceHolder
 DUP
 ALOAD 3
 INVOKEVIRTUAL shark/FieldValuesReader.readId ()J
 INVOKESPECIAL shark/ValueHolder$ReferenceHolder.<init> (J)V
 ASTORE 11
L23
```

Reply:
>  I'm not sure how expensive

Then that's the answer ;) . Favor expression over optimization unless you can prove the optimization is worth the loss in expression